arangod/Cluster/AgencyCache.cpp

```diff
@@ -185,110 +184,119 @@ void AgencyCache::run() {
+                TRI_ASSERT(rs.hasKey("readDB"));
+                std::lock_guard g(_storeLock);
+                LOG_TOPIC("4579f", TRACE, Logger::CLUSTER) <<
+                  "Fresh start: overwriting agency cache with " << rs.toJson();
+                _readDB = rs;                  // overwrite
+                _commitIndex = commitIndex;
               }
-            } else {
-              TRI_ASSERT(rs.hasKey("readDB"));
-              std::lock_guard g(_storeLock);
-              LOG_TOPIC("4579f", TRACE, Logger::CLUSTER) <<
-                "Fresh start: overwriting agency cache with " << rs.toJson();
-              _readDB = rs;                  // overwrite
-              _commitIndex = commitIndex;
-            }
-            triggerWaiting(commitIndex);
-            if (firstIndex > 0) {
-              if (!toCall.empty()) {
-                invokeCallbacks(toCall);
+              triggerWaiting(commitIndex);
+              if (firstIndex > 0) {
+                if (!toCall.empty()) {
+                  invokeCallbacks(toCall);
+                }
+              } else {
+                invokeAllCallbacks();
               }
             } else {
-              invokeAllCallbacks();
+              if (wait <= 1.9) {
+                wait += 0.1;
+              }
+              LOG_TOPIC("9a93e", DEBUG, Logger::CLUSTER) <<
+                "Failed to get poll result from agency.";
             }
-          } else {
+            return futures::makeFuture();
+          })
+        .thenError<VPackException>(
+          [&wait](VPackException const& e) {
+            LOG_TOPIC("9a9f3", ERR, Logger::CLUSTER) <<
+              "Failed to parse poll result from agency: " << e.what();
             if (wait <= 1.9) {
               wait += 0.1;
             }
-            LOG_TOPIC("9a93e", DEBUG, Logger::CLUSTER) <<
-              "Failed to get poll result from agency.";
-          }
-          return futures::makeFuture();
-        })
-      .thenError<VPackException>(
-        [&wait](VPackException const& e) {
-          LOG_TOPIC("9a9f3", ERR, Logger::CLUSTER) <<
-            "Failed to parse poll result from agency: " << e.what();
-          if (wait <= 1.9) {
-            wait += 0.1;
-          }
-        })
-      .thenError<std::exception>(
-        [&wait](std::exception const& e) {
-          LOG_TOPIC("9a9e3", ERR, Logger::CLUSTER) <<
-            "Failed to get poll result from agency: " << e.what();
-          if (wait <= 1.9) {
-            wait += 0.1;
-          }
-        });
-    ret.wait();
-
+          })
+        .thenError<std::exception>(
+          [&wait](std::exception const& e) {
+            LOG_TOPIC("9a9e3", ERR, Logger::CLUSTER) <<
+              "Failed to get poll result from agency: " << e.what();
+            if (wait <= 1.9) {
+              wait += 0.1;
+            }
+          });
+
+      ret.wait();
+    } catch (std::exception const& e) {
+      LOG_TOPIC("544da", ERR, Logger::CLUSTER) <<
```

Comment:
A very good idea, let's also increase the `wait` backoff in this case to avoid excessive log spamming in the unlikely case that a repeated exception is hit. We will still see spamming, but only every 1.9 seconds, which is more bearable than a potentially tight loop with wait == 0. I suggest to do it in exactly the same way as in the `thenError` case.

Reply:
definitely needed. busy loop else

Reply:
done